File: java/AtomAlias.java
/*
 *  Copyright (c) 2014-2015 Bruce Twite
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *	Container as an atomic element.
 */
package yaga.core;

import yaga.core.exceptions.YagaException;
import yaga.core.ristic.AliasRistic;
import yaga.core.ristic.Ristic;

/**
 *
 * @author Bruce
 */
public class AtomAlias extends AtomicList
{
	public AtomAlias(Elements e, ParserPoint point)
		{ super(point); _elements = e; }
	public AtomAlias(List e, ParserPoint point)
		{ super(point); _elements = Elements.make(e); }

	private final Elements _elements;

	@Override
	public List reduce(Context ctxt)
		throws YagaException
	{ 
		if (_elements.isEmpty())
			return (Lists.nil(_point));
		if (_elements.isSingle())
			return (_elements.reduce(ctxt, 0));
		return (Lists.newData(_elements.reduce(ctxt), _point));
	}
	@Override
	public List reduce(Context ctxt, Frame.Reference r)
		throws YagaException
		{ return (r.frame().dispatch(ctxt, () -> reduce(ctxt))); }
	
	@Override
	public List dealias(Context ctxt)
		throws YagaException
	{ 
		if (_elements.isEmpty())
			return (Lists.nil(_point));
		if (_elements.isSingle())
			return (_elements.dealias(ctxt, 0));
		return (Lists.newData(_elements.dealias(ctxt), _point));
	}
	
	@Override
	public List asFrameReference(Context ctxt, Frame frame)
	{
		// Defaults to a Frame reference.
		// May add more discerning detection of variable references in the
		// future.
		return (new Frame.Reference(frame, this));
	}
	
	@Override
	public Symbol getNameSymbol()
	{ 
		if (_elements.isSingle())
			return (_elements.element(0).getNameSymbol()); 
		return (null);
	}

	@Override
	public Name getNameType(Context ctxt) 
		throws YagaException
	{ 
		if (_elements.isSingle())
			return (_elements.element(0).getNameType(ctxt)); 
		return (null);
	}

	@Override
	public Symbol reduceSymbolType(Context ctxt) 
		throws YagaException
	{ 
		if (_elements.isSingle())
			return (_elements.element(0).reduceSymbolType(ctxt)); 
		return (null);
	}

	@Override
	public boolean isBound()
		{ return (_elements.areBound()); }

	@Override
	public boolean isTrivial()
		{ return (_elements.areTrivial()); }

	@Override
	public boolean isReducible()
		{ return (true); }

	@Override
	public boolean hasVariables()
		{ return (_elements.hasVariables()); }

	@Override
	public boolean isAlias()			{ return (true); }

	@Override
	public void print(StringBuilder sb) 
	{
		sb.append("(("); Ristic.risticClassName(sb, AliasRistic.class).append(") ");
		Lists.printElements(sb, _elements);
		sb.append(")");
	}

	@Override
	public void xprint(StringBuilder sb) 
	{
		sb.append("(("); Ristic.risticClassName(sb, AliasRistic.class).append(") ");
		Lists.xprintElements(sb, _elements);
		sb.append(")");
	}
	
	@Override
	public List add(Context ctxt, List e)
		throws YagaException
		{ return (e.rAdd(ctxt, this)); }
	@Override
	public List sub(Context ctxt, List e)
		throws YagaException
		{ return (e.rSub(ctxt, this)); }
	@Override
	public List mul(Context ctxt, List e)
		throws YagaException
		{ return (e.rMul(ctxt, this)); }
	@Override
	public List div(Context ctxt, List e)
		throws YagaException
		{ return (e.rDiv(ctxt, this)); }
	@Override
	public List rem(Context ctxt, List e)
		throws YagaException
		{ return (e.rRem(ctxt, this)); }
}

File: AtomAlias.js
/** 
 *  AtomAlias : @file
 *
 *	Container as an atomic element.
 */

'use strict';

var yc, _prot;
module.exports = {
	new: _newAtomAlias,
	Initialise: _init,
};
Object.freeze(module.exports);


function _newAtomAlias(e, point) {
	let o = Object.create(_prot);
	o.setParserPoint(point);
	if (e.isList) e = yc.Elements.make(e);
	o._elements = e;
	return (o);
}

function _init(y) {
	if (yc) return;
	yc = y;
	let prot = {
		typeName: 'AtomAlias',
		reduce(ctxt) {
			let _elements = this._elements,
				_point = this._point;
			if (_elements.isEmpty())
				return (yc.Lists.nil(_point));
			if (_elements.isSingle())
				return (_elements.reduce(ctxt, 0));
			return (yc.Lists.newData(_elements.reduce(ctxt), _point));
		},
		reduceFrame(ctxt, r) {
			return (r.frame().dispatch(ctxt, () => this.reduce(ctxt)));
		},
		dealias(ctxt) {
			let _elements = this._elements,
				_point = this._point;
			if (_elements.isEmpty())
				return (yc.Lists.nil(_point));
			if (_elements.isSingle())
				return (_elements.dealias(ctxt, 0));
			return (yc.Lists.newData(_elements.dealias(ctxt), _point));
		},
		asFrameReference(ctxt, frame) {
			// Defaults to a Frame reference.
			// May add more discerning detection of variable references in the
			// future.
			return (yc.Frame.Reference.new(frame, this));
		},
		getNameSymbol() {
			if (this._elements.isSingle())
				return (this._elements.element(0).getNameSymbol());
			return (null);
		},
		getNameType(ctxt) {
			if (this._elements.isSingle())
				return (this._elements.element(0).getNameType(ctxt));
			return (null);
		},
		reduceSymbolType(ctxt) {
			if (this._elements.isSingle())
				return (this._elements.element(0).reduceSymbolType(ctxt));
			return (null);
		},
		isBound() {
			return (this._elements.areBound());
		},
		isTrivial() {
			return (this._elements.areTrivial());
		},
		isReducible: () => true,
		hasVariables() {
			return (this._elements.hasVariables());
		},
		isAlias: () => true,
		print(sb) {
			sb.append("((");
			yc.Ristic.risticClassName(sb, yc.AliasRistic).append(") ");
			yc.Lists.printElements(sb, this._elements);
			sb.append(")");
		},
		xprint(sb) {
			sb.append("((");
			yc.Ristic.risticClassName(sb, yc.AliasRistic).append(") ");
			yc.Lists.xprintElements(sb, this._elements);
			sb.append(")");
		},
		add(ctxt, e) {
			return (e.rAdd(ctxt, this));
		},
		sub(ctxt, e) {
			return (e.rSub(ctxt, this));
		},
		mul(ctxt, e) {
			return (e.rMul(ctxt, this));
		},
		div(ctxt, e) {
			return (e.rDiv(ctxt, this));
		},
		rem(ctxt, e) {
			return (e.rRem(ctxt, this));
		},
	};
	_prot = Object.assign(Object.create(yc.List.prototype), prot);
}